v0.1.0

asqarslanov opened the pull request.

asqarslanov committed:
I've done too much not to commit

asqarslanov committed:
refactor: create a Russian locale file

asqarslanov committed:
refactor: add data structures

asqarslanov committed:
fix: validate `gpa` inside `calculateScholarship`

asqarslanov committed:
refactor: remove ru.ftl

evermake commented on src/bot.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+bot.on("message:text", async (ctx) => {
+  const gpa = parseFloat(ctx.message.text);
+
+  try {
+    const scholarship = calculateScholarship(gpa);
+    await ctx.reply(scholarship.toString());
+  } catch (_) {
+    ctx.reply("Invalid input!");
+  }
```
Remove try/catch and validate user input explicitly

asqarslanov replied:
Then the program will crash in case of wrong input.

evermake replied:
You should validate `gpa` and notify a user, if it's invalid. If error will be thrown after validation, then there is actually an error in program and it should throw.

asqarslanov replied:
But if I don't validate the value in the function, won't it be unsafe? If the function receives incorrect data, it should either throw an error or return `null`/`undefined`/whatever. I cannot just assume that the `gpa` argument is always in the valid range. I could use another type for this argument, but it's pretty tough in TS to specify the range for a value.

dantetemplar replied:
you can use some specific exception instead of wildcart. In Python, f.e. ValueError used for any validation-related stuff.

> But if I don't validate the value in the function, won't it be unsafe? If the function receives incorrect data, it should either throw an error or return null/undefined/whatever. I cannot just assume that the `gpa` argument is always in the valid range. I could use another type for this argument, but it's pretty tough in TS to specify the range for the value.

asqarslanov replied:
If I understand you correctly, that's exactly what I'm doing. The function throws RangeError if `gpa` is not in the range [2, 5]. Then each time calling this function I catch this error.

dantetemplar replied:
So, maybe specifying Exception type in catch statement will be appropriate? 

```TypeScript
try {
    // Code
} catch (e: any) {
    if (e instanceof YourException){
        // code that runs for this exception.
    } else if (e instanceof YourException2){
        // code that runs for this exception.
    } else {
       // Throw Runtime Error or something like that, I really don't know TS...
    }
}
```

dantetemplar replied:
Also you can check how Artem has solved this: 
https://github.com/one-zero-eight/InNoHassle-Website/blob/main/components/ScholarshipCalculator.tsx

asqarslanov replied:
Actually, this seems to be the solution!

evermake committed:
improve config loading, add `zod`

evermake committed:
update `.gitignore`, add VSCode recommended setup

evermake committed:
minor refactoring

asqarslanov committed:
chore: helix lsp configuration

asqarslanov committed:
fix: start the bot without confirmation

asqarslanov committed:
style: remove extra indentation

asqarslanov committed:
feat: propagate errors to users

asqarslanov committed:
feat: store sessions as files

asqarslanov committed:
chore: deno.lock grammy_storages

asqarslanov committed:
doc: default dir name of sessions

asqarslanov committed:
refactor(deno): update dependencies 10/18/2023

asqarslanov committed:
refactor: type_hints.ts

asqarslanov committed:
BREAKING CHANGE: not my best commit

evermake commented:
я бы не использовал плагин conversations

asqarslanov commented:
> я бы не использовал плагин conversations

Тогда в `SessionData` хранить ветку диалога?

evermake commented:
В сессии хранить состояние, да. И на его основе можно роутить, используя Composer.route(), или просто в middleware писать логику.

asqarslanov commented:
> В сессии хранить состояние, да. И на его основе можно роутить, используя Composer.route(), или просто в middleware писать логику.

Возможно, я неправильно понял, но сами авторы фреймворка [не рекомендуют](https://grammy.dev/plugins/router#combining-routers-with-sessions) использование роутеров:

_Note that a much better solution is to use the [conversations plugin](https://grammy.dev/plugins/conversations). The remainder of this page is obsolete since that plugin was created. We will keep this page as a reference for those who `used the router for forms._

evermake commented:
Потому что они написали удобный плагин и рекомендуют использовать его, но на практике у этого плагина есть проблемы

asqarslanov committed:
feat: initial conversation (not done)

evermake commented:
@asqarslanov , код ещё не полностью готов, как я понял. Напиши мне, чтобы я ревью сделал, как ± готово будет.

asqarslanov committed:
feat: load schedules on fly

asqarslanov committed:
refactor(config): async parse

asqarslanov committed:
style(config): envPath

asqarslanov committed:
refactor: load schedules

asqarslanov committed:
refactor(config): schedule loading code scheme

asqarslanov committed:
chore(helix): new lsp config

asqarslanov committed:
chore(helix): full deno lsp

asqarslanov committed:
chore(helix): md lsp roots

asqarslanov committed:
feat: load schedules every minute

asqarslanov committed:
refactor: Config -> ZConfig

asqarslanov committed:
refactor: function imports with the modules

asqarslanov committed:
refactor: getScholarship() -> scholarship.get()

asqarslanov committed:
feat: scholarship module (not done yet)

evermake commented on src/schedules.ts:
```diff
@@ -19,8 +21,16 @@ const Schedules = z.object({
+  return await fetch(url)
+    .then(async (s) => await s.json())
+    .then(async (s) => await ZSchedules.parseAsync(s));
+}
```
Async callbacks in `Promise.then` — it's a bad practice. If you're writing async function, **just use `await`s**. Like this:

```js
export async function getSchedules(url) {
  const response = await fetch(url)
  const data = await response.json()
  const schedules = await ZSchedules.parseAsync(data)
  return schedules
}
```

or if you want more concise solution, you probably don't need an `async`:

```js
const getSchedules = (url) => fetch(url)
  .then((resp) => resp.json())
  .then((data) => ZSchedules.parseAsync(data))
```

---

Also it seems like `ZSchedules.parseAsync` is unnecessary here and you just need to use synchronous `parse`.

asqarslanov committed:
refactor(schedules): getting rid of a bad practice

asqarslanov committed:
feat(scholarship): gpa parsing

asqarslanov committed:
refactor: schedules -> event groups

asqarslanov committed:
style: deno.jsonc - kebab-case

asqarslanov committed:
chore: new dependency (ical.js)

asqarslanov committed:
refactor: separate the code into modules

asqarslanov committed:
refactor: converations composer explicit name

asqarslanov committed:
refactor(session): state to const enum

asqarslanov committed:
refactor: re-exports

asqarslanov committed:
style(imports): coversations/scholarship

asqarslanov committed:
refactor(session): split state into conversation and context

asqarslanov committed:
feat: information button

asqarslanov committed:
cringe: print id

asqarslanov committed:
refactor(labels): separate module

asqarslanov committed:
fix(locales): grammar, recommend doing

asqarslanov committed:
fix: `ical` -> `ical-js`

asqarslanov committed:
refactor(session): split into different files

asqarslanov committed:
feat(session): user's course

asqarslanov committed:
refacor(scholarship): move to `conversations`

asqarslanov committed:
fix(session): export types

asqarslanov committed:
feat(scholarship): calculating logic

asqarslanov committed:
fix(session): store enums as consts

asqarslanov committed:
fix: import types everywhere

asqarslanov committed:
refactor(locales): emoji terms, long--kebab--case

asqarslanov committed:
feat: progress bar generation (integration with iu sport)

asqarslanov committed:
docs: `progress_bar.generate()`

asqarslanov committed:
chore: `assert` for probable unit-testing

asqarslanov committed:
fix(locales): url-108-tg "..."

asqarslanov committed:
fix: import `I18n` correctly

evermake commented on locales/en.ftl:
```diff
@@ -1,24 +1,39 @@
+-url--innohassle = https://innohassle.ru/
+-url--one-zero-eight-telegram = "https://t.me/one_zero_eight"
+
+-emoji--check-mark-button = ✅
+-emoji--gear = ⚙️
+-emoji--left-arrow = ⬅️
+-emoji--money-with-wings = 💸
+-emoji--no-entry = ⛔️
+-emoji--ok-hand = 👌
+-emoji--open-book = 📖
+-emoji--right-arrow = ➡️
+-emoji--speech-balloon = 💬
+-emoji--tear-off-calendar = 📆
+-emoji--white-question-mark = ❔
```
These ones only add clutter and decrease readability of the messages.

asqarslanov replied:
There’s actually a technical reason for this. When using “pure” emoji symbols, the file may not be readable in some editors / with some fonts. Emojis aren’t that widespread in programming and monospaced environments.

evermake replied:
But you still use them in the same file... What's the difference then?

evermake commented on src/conversations/scholarship/scholarship_logic.ts:
```diff
@@ -20,31 +22,39 @@ export async function parseGpa(str: string): Promise<number> {
+}
 
-  let s = M_MIN + (M_MAX - M_MIN) * ((gpa - 2) / 3) ** 2.5;
+function apply_formula(gpa: number, m_min: number, m_max: number): number {
```
We use camelCase for function names in JS/TS. Also I think it would be better to move these functions from `conversations` directory to some `utils`, for example.

evermake replied:
Or even better create a directory for business logic.

asqarslanov replied:
1) Yes, I know about camelCase, my bad. Gotta fix it.

2) I decided to structure the project as a tree of modules. These functions are located at `conversation/scholarship/scholarship_logic` since they are only used for the `scholarship` conversation. Using something like `utils` would make the structure less clear.

evermake replied:
Then why do you create directories and modules with the same name (`scholarship` and `scholarship.ts`)?

asqarslanov committed:
docs: side effects, clearer explanation

evermake commented on src/conversations/settings.ts:
```diff
@@ -0,0 +1,17 @@
+
+export const composer = new Composer<MyContext>();
+
+composer.callbackQuery(Button.MainMenuSettings, async (ctx) => {
+  await ctx.deleteMessage();
+
+  const inlineKeyboard = new InlineKeyboard()
+    .text(ctx.t(Button.BackToMainMenu), Button.BackToMainMenu);
+
+  await ctx.reply(ctx.t(Message.Settings), {
+    reply_markup: inlineKeyboard,
+  });
+});
```
In Telegram we update messages with the new keyboard instead of deleting them and sending new.

asqarslanov replied:
I believe, if I change the contents of a message, it will appear to the user that the message was edited, which doesn’t look professional.

evermake replied:
It won't. All bots work in this way, including BotFather.

asqarslanov replied:
Oh, indeed. I’ll look into that.

evermake commented on src/conversations/scholarship/scholarship_logic.ts:
```diff
@@ -0,0 +1,60 @@
+
+import { Course, Year } from "@/bot.ts";
+
+export async function parseGpa(str: string): Promise<number> {
```
There is no reason for this function to be asynchronous.

asqarslanov committed:
Merge branch 'v0.1.0' of github.com:one-zero-eight/InNoHassle-Bot into v0.1.0

asqarslanov committed:
fix(scholarship): style, api, non-async

asqarslanov committed:
refactor(scholarship): only accept strings when parsing gpas

evermake commented on src/conversations/scholarship.ts:
```diff
@@ -0,0 +1,41 @@
+  const grades = ctx.message?.text;
+
+  if (grades === undefined) {
+    return onMessage(ctx);
```
Infinite recursion?

asqarslanov replied:
It’s more like a temporal solution, but, basically, yes. The user will be notified about their error clearly in the future, and there will be a back button to stop the conversation.

evermake replied:
ok

asqarslanov committed:
refactor: replace `@/` with `~/`

This better resembles UNIX conventions with `~` being the home directory.
Also, this style is used in other web frameworks (e.g. SolidStart).

asqarslanov committed:
fix(locales): start--double-kebab-case

asqarslanov committed:
chore!: deno tasks `start` and `run`

asqarslanov committed:
refactor: get use of `export default`

asqarslanov committed:
docs(readme): description, stack, style guides, and local development

asqarslanov committed:
docs(readme): screenshot

asqarslanov committed:
chore: bot assets

asqarslanov committed:
fix(conversations): update messages instead of resending them

asqarslanov committed:
docs(readme): fix screenshot size

asqarslanov committed:
docs(readme): minor changes to module organization and linking

asqarslanov committed:
refactor: add a template conversation

asqarslanov committed:
style(imports): `type` inside of braces

asqarslanov committed:
style!: rename schema keys to camelCase

asqarslanov committed:
refactor: no default parameter values in functioins

asqarslanov committed:
style: conventional zod schemas naming

asqarslanov committed:
screw conventional commits

The development process goes much faster without worrying about nice commit descriptions. It's my repo, I do what I want. This is a huge commit, but I don't care anymore.

asqarslanov committed:
no reexporting

evermake commented on deno.jsonc:
```diff
@@ -10,6 +10,7 @@
   },
   "tasks": {
-    "start": "deno run --allow-env --allow-net --allow-read --allow-write src/bot.ts"
+    "run": "deno run --allow-env --allow-net --allow-read --allow-write src/bot.ts",
+    "start": "trash sessions/ || rm --force --recursive sessions/; deno task run"
```
From name of the tasks there is no difference. I'd remove the `start` task and mention somewhere about the `sessions/` directory.

asqarslanov replied:
Okay. I guess, `run`, `clear`, and `run-clear` will be fine.

evermake commented on README.md:
Great documentation! However, I'd shorten the `Local Development` to 2-3 lines of text, its toooo verbose. I hope IU students know how to clone a repo or how to open directory in CLI, otherwise, the contribution is impossible.

asqarslanov replied:
Good point 😅

evermake commented on README.md:
```diff
@@ -1 +1,123 @@
+- [`src/labels`](src/labels) — most string literals are wrapped into enums here
+  for the sake of maintainability and IDE suggestions.
+
+## Local Deveolpment
```
typo

evermake commented on src/config.ts:
```diff
@@ -8,4 +8,5 @@ await dotenv.load({
 });
 
-export const config = await ZConfig.parseAsync(Deno.env.toObject());
+const config = await ZConfig.parseAsync(Deno.env.toObject());
```
Just `.parse`, no need for `parseAsync`.

asqarslanov committed:
Thank you, Vladislav!

asqarslanov committed:
in sync with readme

asqarslanov committed:
sport button main menu

asqarslanov committed:
scholarship finally done

asqarslanov committed:
inh id string

asqarslanov committed:
message id should always exist

asqarslanov committed:
embrace algebraic data types! (sorta)

asqarslanov committed:
locale error handling formatting

asqarslanov committed:
no info conversation so far

asqarslanov committed:
bot restart

asqarslanov committed:
perfect scholarship
